windows-installer/openssl.nsi

```diff
@@ -1,111 +1,144 @@
+NAME "${PRODUCT_NAME} ${VERSION}"
+
+ShowInstDetails show
+
+
+Var DataDir
+Var ModDir
+
+Function .onInit
+    StrCpy $INSTDIR "C:\Program Files\openssl-${MAJOR}.${MINOR}"
+FunctionEnd
+
+# This section is run if installation of 32 bit binaries are selected
+
+!ifdef BUILD64
+# This section is run if installation of the 64 bit binaries are selectd
+SectionGroup "64 Bit Installation"
+    Section "64 Bit Binaries"
+        SetOutPath $INSTDIR\x64\lib
+        File /r "${BUILD64}\Program Files\OpenSSL\lib\"
+        SetOutPath $INSTDIR\x64\bin
+        File /r "${BUILD64}\Program Files\OpenSSL\bin\"
+        SetOutPath "$INSTDIR\x64\Common Files"
+        File /r "${BUILD64}\Program Files\Common Files\"
+    SectionEnd
+    Section "x64 Development Headers"
+        SetOutPath $INSTDIR\x64\include
+        File /r "${BUILD64}\Program Files\OpenSSL\include\"
+    SectionEnd
+SectionGroupEnd
+!endif
+
+!ifdef BUILD32
+# This section is run if installation of the 64 bit binaries are selectd
+SectionGroup "32 Bit Installation"
+    Section "32 Bit Binaries"
+        SetOutPath $INSTDIR\x32\lib
+        File /r "${BUILD32}\Program Files (x86)\OpenSSL\lib\"
+        SetOutPath $INSTDIR\x32\bin
+        File /r "${BUILD32}\Program Files (x86)\OpenSSL\bin\"
+        SetOutPath "$INSTDIR\x64\Common Files"
+        File /r "${BUILD32}\Program Files (x86)\Common Files\"
+    SectionEnd
+    Section "x32 Development Headers"
+        SetOutPath $INSTDIR\x32\include
+        File /r "${BUILD32}\Program Files (x86)\OpenSSL\include\"
+    SectionEnd
+SectionGroupEnd
+!endif
+
+!ifdef BUILD64
+Section "Documentation"
+    SetOutPath $INSTDIR\html
+    File /r "${BUILD64}\Program Files\OpenSSL\html\"
+SectionEnd
+!endif
+
+# Always install the uninstaller and set a registry key
+Section
+    WriteUninstaller $INSTDIR\uninstall.exe
+SectionEnd
+
+!define env_hklm 'HKLM "SYSTEM\CurrentControlSet\Control\Session Manager\Environment"'
+!define openssl_hklm 'HKLM "SOFTWARE\OpenSSL-${MAJOR}.${MINOR}-${CTX}"'
+
+# This is run on uninstall
+Section "Uninstall"
+    RMDIR /r $INSTDIR
+    DeleteRegValue ${openssl_hklm} OPENSSLDIR
+    DeleteRegValue ${openssl_hklm} MODULESDIR
+    DeleteRegValue ${openssl_hklm} ENGINESDIR
+    SendMessage ${HWND_BROADCAST} ${WM_WININICHANGE} 0 "STR:Environment" /TIMEOUT=5000
+SectionEnd
+
+!insertmacro MUI_PAGE_WELCOME
+
+!insertmacro MUI_PAGE_LICENSE ${LICENSE_FILE}
+
+Function CheckRunUninstaller
+!ifdef BUILD64
+    StrCpy $DataDir "$INSTDIR\x64\Common Files\SSL"
+    StrCpy $ModDir  "$INSTDIR\x64\lib\ossl-modules"
+!else
+    StrCpy $DataDir "$INSTDIR\x32\Common Files\SSL"
+    StrCpy $ModDir  "$INSTDIR\x32\lib\ossl-modules"
+!endif
+    ifFileExists $INSTDIR\uninstall.exe 0 +2
+        ExecWait "$INSTDIR\uninstall.exe /S _?=$INSTDIR"
+
+    WriteRegExpandStr ${openssl_hklm} OPENSSLDIR "$DataDir"
+    WriteRegExpandStr ${openssl_hklm} ENGINESDIR "$ModDir"
+    WriteRegExpandStr ${openssl_hklm} MODULESDIR "$ModDir"
```

Comment:
This looks like we are inserting these keys on uninstall which I don't understand. Can you explain what is happening here?

Reply:
CheckRunUninstaller is run universally on exit from the DIRECTORYPAGE_TEXT_DESTINATION panel.  It actually only runs the uninstaller if it exists. see:
```
ifFileExists $INSTDIR\uninstall.exe 0 +2
        ExecWait "$INSTDIR\uninstall.exe /S _?=$INSTDIR"
```

But I agree, its a bit fuzzy.  I can refactor that